### ISSUE TITLE
Errors with nanoid v3 when using jest and jsdom

### ISSUE THREAD

**Issue by anomiex** [ai/nanoid#462]
When running tests using jest and jest-environment-jsdom, version 3.3.7 of this package (which is still supposed to be commonjs-compatible) causes a failure:
```
 FAIL  ./test.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /tmp/test/node_modules/nanoid/index.browser.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){import { urlAlphabet } from './url-alphabet/index.js'
                                                                                      ^^^^^^

    SyntaxError: Cannot use import statement outside a module

      3 |  */
      4 |
    > 5 | const nanoid = require( 'nanoid' );
        |                ^
      6 |
      7 | test( 'it works', () => {
      8 |     const element = document.createElement('div');

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1505:14)
      at Object.require (test.js:5:16)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.298 s
```

The problem is that, with jest-environment-jsdom, the conditions available are "browser", "require" (not "import"), and "default", but nanoid v3's `exports` assumes "browser" always comes with "import".

A detailed analysis of a similar problem in another package, along with potential workarounds, may be found at https://github.com/microsoft/accessibility-insights-web/pull/5421#issuecomment-1109168149.

### Reproduction

1. Create an empty directory for the test.
2. Run `npm add jest jest-environment-jsdom nanoid@^3`
3. Create the following file named `test.js`:
   ```js
   /**
    * @jest-environment jsdom
    */
   
   const nanoid = require( 'nanoid' );
   
   test( 'it works', () => {
       const element = document.createElement('div');
       expect( element ).not.toBeNull();
   } );
   ```
5. Run `npm exec jest test.js`
6. See error

**Comment by ai** [comment]
> The problem is that, with jest-environment-jsdom, the conditions available are "browser", "require" (not "import"), and "default", but nanoid v3's exports assumes "browser" always comes with "import".

What we can do on our side if the other module works with `package.exports` in a wrong way?

**Comment by cmdcolin** [comment]
similar thread here https://github.com/ai/nanoid/issues/439

**Comment by anomiex** [comment]
In what way is jest with jsdom working with `package.exports` in a "wrong" way here? It's specifying "browser" and "require" (and not "import"), which is exactly what it wants to receive.

**Comment by ai** [comment]
> It's specifying "browser" and "require" (and not "import"), which is exactly what it wants to receive.

`browser` version of library could not be compatible for Node.js environment. How do you plan to test code if it was created for different environment?

**Comment by ai** [comment]
Here is current `package.exports` of v3:

```json
  "exports": {
    ".": {
      "browser": "./index.browser.js",
      "require": {
        "types": "./index.d.cts",
        "default": "./index.cjs"
      },
      "import": {
        "types": "./index.d.ts",
        "default": "./index.js"
      },
      "default": "./index.js"
    }
```

What changes do you suggest there to fix Jest?

**Comment by anomiex** [comment]
jsdom is supposed to provide a browser-like environment. If it's not browser-like enough, that would be a problem with jsdom. OTOH, if a package assumes "browser" means "import", that's not jsdom's fault.

> What changes do you suggest there to fix Jest?

You already appear to have an `index.browser.cjs` file available in the published package, so something like this would work.
```
  "exports": {
    ".": {
      "browser": {
        "require": "./index.browser.cjs",
        "default": "./index.browser.js",
      },
      "require": {
        "types": "./index.d.cts",
        "default": "./index.cjs"
      },
      "import": {
        "types": "./index.d.ts",
        "default": "./index.js"
      },
      "default": "./index.js"
    }
```
Or if you're worried that some browser might supply both "require" and "import", you could do like this to prefer the esm version in that case
```
      "browser": {
        "import": "./index.browser.js",
        "require": "./index.browser.cjs",
        "default": "./index.browser.js",
      },
```
Or you could switch it up, put "browser" inside the existing "require" and "import" blocks:
```
  "exports": {
    ".": {
      "require": {
        "types": "./index.d.cts",
        "browser": "./index.browser.cjs",
        "default": "./index.cjs"
      },
      "import": {
        "types": "./index.d.ts",
        "browser": "./index.browser.js",
        "default": "./index.js"
      },
      "browser": "./index.browser.js",
      "default": "./index.js"
    }
```

**Comment by ai** [comment]
But is it a good idea to use browser’s version in Node.js? Does Jest emulate `crypto` for Node.js <20?

**Comment by anomiex** [comment]
I'd say that if someone wants to try it, as indicated by passing the "browser" condition, it's their problem if it doesn't work due to a missing Web Crypto API or whatever.

In my case we're running the tests in Node 20 already anyway. But if we were still testing with an earlier version, I'd take it as my responsibility to load an appropriate mock/polyfill in the Jest config to make it work.

**Comment by ai** [comment]
Taking `require` was just work. We have a special Nose.js version exactly to work without polyfill.

OK, can you send PR to `dual-publish` to add `browser.require` if it already produces CJS files for browser?

**Comment by MikeMulyukin1983** [comment]
any solution for fix this?

**Comment by AdiMarianMutu** [comment]
I get the same error, however related to `import { webcrypto as crypto } from 'node:crypto'`.

Current `nanoid` version: `5.0.7`

Is there any solution to this?

**Comment by ai** [comment]
@AdiMarianMutu you have a different issue. You need to update your project to ESM to use 4.0 or 5.0.

**Comment by good-idea** [comment]
If this is blocking anyone, you can make a simple jest mock to get around it:

```
// in jest setup config

jest.mock('nanoid', () => {
  return {
    nanoid: () => Math.random().toString(),
  };
});
```

**Comment by jhackett1** [comment]
have removed from our codebase for this reason sadly :(

**Comment by rujorgensen** [comment]
I switched to the native 'crypto.randomUUID()' implementation in the end. One dependency less, haven't been happier.

**Comment by ppamorim** [comment]
@ai Any solution for that in mind? We will keep using v3 while this doesn't change. Sadly we can't get the improvements done for the node:18.

**Comment by ai** [comment]
@ppamorim sorry, there is no good solution. Jest is just broken with ESM because it hacked Node.js loading mechanism.

All options have a lot of compromises.

**Comment by anomiex** [comment]
The breakage here has nothing to do with jest "hack[ing] the Node.js loading mechanism", it's that this package assumes that "browser" always means "import" despite the loading request specifically indicating "require" (i.e. commonjs) rather than "import" (i.e. esm).

I told you [back in January](https://github.com/ai/nanoid/issues/462#issuecomment-1875818424) how to fix it in v3, but I can't figure out your custom build system to work out [how to fix it for you](https://github.com/ai/nanoid/issues/462#issuecomment-1876087266).

Although it seems like ppamorim may be asking about something else, since this issue is specific to v3 so "We will keep using v3 while this doesn't change" makes little sense in relation to _this_ issue.

**Comment by ai** [comment]
> it's that this package assumes that "browser" always means "import" despite the loading request specifically indicating "require" (i.e. commonjs) rather than "import" (i.e. esm).

Jest is broken because it is using browser version on Node.js environment.

We provide CJS package for Node.js, but Jest hacked `require` to use browser version which lead to issue with RNG too.

**Comment by anomiex** [comment]
We already went through that argument above.

[me](https://github.com/ai/nanoid/issues/462#issuecomment-1875769491):
> In what way is jest with jsdom working with `package.exports` in a "wrong" way here? It's specifying "browser" and "require" (and not "import"), which is exactly what it wants to receive.


[you](https://github.com/ai/nanoid/issues/462#issuecomment-1875828909):
> But is it a good idea to use browser’s version in Node.js? Does Jest emulate `crypto` for Node.js <20?
> 
> We didn’t use `.browser.cjs` especially for that reason since it didn’t work in Node.js.


[me](https://github.com/ai/nanoid/issues/462#issuecomment-1876083106):
> I'd say that if someone wants to try it, as indicated by passing the "browser" condition, it's their problem if it doesn't work due to a missing Web Crypto API or whatever.
> 
> In my case we're running the tests in Node 20 already anyway. But if we were still testing with an earlier version, I'd take it as my responsibility to load an appropriate mock/polyfill in the Jest config to make it work.

**Comment by ai** [comment]
Browser version in Node.js environment is still non-standard Node.js behavior and this is why for me, it is a hack.

**Comment by Grohden** [comment]
if it helps with anything, I couldn't mock the module due to having a `requireActual` in a dep that used `nanoid` (at least thats why I think the mock didn't work)

However by letting jest transform `nanoid` it works fine for me:

```js
// jest.config.js

{
 transformIgnorePatterns: [
    '<rootDir>/node_modules/(?!nanoid)',
  ],
}
```

**Comment by huksley** [comment]
This `transformIgnorePatterns: ['<rootDir>/node_modules/(?!nanoid)']` have worked for me, but only after I fixed my configuration, produced with combination of Jest + NextJS. I use (NextJS 14.2) and it adds this pattern to `transformIgnorePatterns: ["/node_modules/"]`. I had to remove it.

```typescript
const exportConfig = async () => {
  const config = await createJestConfig(customJestConfig)();
  return {
    ...config,
    transformIgnorePatterns: config.transformIgnorePatterns?.filter(
      // NextJS adds this pattern and all your patterns does not work afterwards
      pattern => pattern !== "/node_modules/"
    )
  };
};

export default exportConfig;
```

**Comment by Grohden** [comment]
@huksley for next they kinda override the transform ignore patterns (it I was VERY happy to discover that :rage1:), you might need to handle that in `transpilePackages`

https://nextjs.org/docs/app/api-reference/config/next-config-js/transpilePackages

**Comment by timkindberg** [comment]
We had the same problem with the [uuid package](https://github.com/uuidjs/uuid/pull/616). We fixed it like so (which also fixes this module) in the jest.config.js:

```js
moduleNameMapper: {
    // This is a workaround for https://github.com/uuidjs/uuid/pull/616
    '^uuid$': require.resolve('uuid'),
    '^nanoid$': require.resolve('nanoid')
  },
```

I'm not sure if the uuid package has found a fix, maybe if they have you can mimic it.

**Comment by naveedehmad** [comment]
yeah, had to remove the package for the exact same thing. `crypto.randomUUID()` it is until it's fixed. can't just mock it in all of test files.

**Comment by jwarykowski** [comment]
Just to note here, I've had a similar experience but within a react-native context. The exports react-native property currently point to an ESM version of the build (`"react-native": "./index.browser.js"`).

As an experiment, if you update the version of the lib in your node modules folder to point to a cjs build, the tests when run with jest will now work. Why wouldn't these point to a cjs build by default?

Note in order to get around this issue I simply added `nanoid` to the `transformIgnorePatterns` as mentioned above.

**Comment by orhanayd** [comment]
@anomiex 
# Why choose nope-id?

**nope-id is a drop-in, nanoid-compatible ID generator — without the module/compatibility headaches, with the ID formats real apps need, and faster. Zero dependencies.**

```js
import { nopeid } from 'nope-id'        // ESM
const { nopeid } = require('nope-id')   // CommonJS
nopeid() // "V1StGXR8_Z5jdHi6B-myT"
```

## Problems you simply won't have

The compatibility issues people keep running into elsewhere — **don't exist with nope-id**, and that's verified, not promised:

- ✅ **CommonJS just works.** `require('nope-id')` returns the API on every Node ≥ 16. No ESM-only surprises, no `ERR_REQUIRE_ESM`, no "downgrade to an older version" workarounds. (cf. nanoid [#365](https://github.com/ai/nanoid/issues/365), where v4 dropped CommonJS.)
- ✅ **Jest + jsdom just works.** A `jest-environment-jsdom` test doing `const { nopeid } = require('nope-id')` passes — no `SyntaxError: Cannot use import statement outside a module`. We actually ran it. (cf. nanoid [#462](https://github.com/ai/nanoid/issues/462).)
- ✅ **Everywhere else, too.** Node (ESM + CJS), browsers, modern bundlers (Vite/webpack/Rollup/esbuild), and React Native — each gets the right build, with no config gymnastics on your side.

We ship a **real dual build** (`index.js` ESM + `index.cjs` CommonJS + a Web-Crypto browser build) with a correctly-ordered `exports` map. That's the whole point: you import it and move on.

## And you get more than nanoid

nanoid gives you one URL-safe random string. nope-id gives you that **plus** everything production actually needs — all tree-shakeable, so you only pay for what you import:

- **Sortable, index-friendly keys:** `uuidv7()`, spec-compliant `ulid()` + `monotonicFactory()`, `snowflake()`
- **Ecosystem compatibility:** `objectId()` (MongoDB), `sqidsFactory()` (reversible short codes)
- **Developer experience:** `defineId('user')` → typed `` `user_${string}` `` IDs; `shortId()`/`slugId()` for human-friendly codes
- **Utilities:** `isValid()`, `isValidUUID()`, `isValidULID()`, `decodeTime()`, collision calculator
- **Security hardening:** CSPRNG, modulo-bias-free sampling, constant-time validation, prototype-pollution-safe alphabets

## Faster, too

Benchmarked against nanoid 5.1.11 (best-of-5 with a fair global warmup — run `npm run benchmark` yourself):

| | nope-id vs nanoid |
|---|---|
| Basic (21) | **~+18%** |
| Small (10) | **~+9%** |
| Large (64) | **~+17%** |
| Custom alphabet | **~+20%** |
| Batch (100) | **~+15%** |

**Wins all 5 core benchmarks** — while shipping far more, with zero dependencies.

## At a glance

| | nope-id | nanoid 5.x |
|---|---|---|
| ESM **+ CommonJS** | ✅ | ⚠️ ESM-only (dropped CJS in v4) |
| Works under Jest + jsdom (`require`) | ✅ verified | ⚠️ known breakage (#462) |
| Sortable IDs (UUIDv7 / ULID / Snowflake) | ✅ | ❌ |
| ObjectId / Sqids / typed IDs / validators | ✅ | ❌ |
| Speed (core benchmarks) | **wins 5/5** | — |
| Dependencies | 0 | 0 |

## Bottom line

If you want nanoid's ergonomics **without** the ESM/CommonJS and Jest/jsdom friction — and with the ID formats, security hardening, and speed production needs — choose nope-id.

```bash
npm install nope-id
```

**Comment by ai** [comment]
Closing it since ecosystem is going over it

**Comment by anomiex** [comment]
> Closing it since ecosystem is going over it

What does this mean?

FYI, I can still reproduce the bug following the instructions in the description.

**Comment by ai** [comment]
@anomiex do you have an issue in normal Jest with Nano ID 5 and ESM ecosystem?

**Comment by anomiex** [comment]
This bug isn't about nanoid v5, it's specifically about nanoid v3.

Nanoid v4+ going ESM-only causes different issues with normal Jest, which are outside the scope of this bug.

**Comment by ai** [comment]
The modern ecosystem has pretty good ESM sorry so we don't need Nano ID 3.x right now. 

I am planning to drop 3.x support around next year. In this case fixing big with Jest hack would not be done.

**Comment by anomiex** [comment]
If you don't want to support v3 anymore, that's your prerogative. You might want to update https://github.com/ai/nanoid/blob/0b695546fd83beaa943fbcaa90af92895dd0ae55/README.md?plain=1#L156-L160 in that case though.
